README.md
[![Build Status](https://github.com/NickleDave/evfuncs/actions/workflows/ci.yml/badge.svg)
[![DOI](https://zenodo.org/badge/158776329.svg)](https://zenodo.org/badge/latestdoi/158776329)
[![PyPI version](https://badge.fury.io/py/evfuncs.svg)](https://badge.fury.io/py/evfuncs)
[![License](https://img.shields.io/badge/License-BSD%203--Clause-blue.svg)](https://opensource.org/licenses/BSD-3-Clause)
# *ev*funcs
Functions for working with files created by EvTAF and the evsonganaly GUI.  
In case you need to work with those files in Python 😊😊😊 (see "Usage" below).

The first work published with data collected using EvTAF and evsonganaly is in this paper:  
Tumer, Evren C., and Michael S. Brainard.  
"Performance variability enables adaptive plasticity of ‘crystallized’adult birdsong."  
Nature 450.7173 (2007): 1240.  
<https://www.nature.com/articles/nature06390>  

These functions are translations to Python of the original functions 
written in MATLAB (copyright Mathworks) by Evren Tumer (shown below).  
<p style="text-align:center;">
<img src="./doc/ev_ev_ev.png" alt="Image of Evren">
</p>

### Installation
#### with `pip`

```console
$ pip install evfuncs
```

#### with `conda`

```console
$ conda install evfuncs -c conda-forge
```


### Usage

The main purpose for developing these functions in Python was to 
work with files of Bengalese finch song in this data repository: 
<https://figshare.com/articles/Bengalese_Finch_song_repository/4805749>

Using `evfuncs` with that repository, you can load the `.cbin` audio files ...
```Python
>>> import evfuncs

>>> rawsong, samp_freq = evfuncs.load_cbin('gy6or6_baseline_230312_0808.138.cbin')
```

... and the annotation in the `.not.mat` files ...
```Python
>>> notmat_dict = evfuncs.load_notmat('gy6or6_baseline_230312_0808.138.cbin')
```
(or, using the `.not.mat` filename directly)
```Python
>>> notmat_dict = evfuncs.load_notmat('gy6or6_baseline_230312_0808.138.not.mat')
```

...and you should be able to reproduce the segmentation of the raw audio files of birdsong
into syllables and silent periods, using the segmenting parameters from a .not.mat file and 
the simple algorithm applied by the SegmentNotes.m function.

```Python
>>> smooth = evfuncs.smooth_data(rawsong, samp_freq)
>>> threshold = notmat_dict['threshold']
>>> min_syl_dur = notmat_dict['min_dur'] / 1000
>>> min_silent_dur = notmat_dict['min_int'] / 1000
>>> onsets, offsets = evfuncs.segment_song(smooth, samp_freq, threshold, min_syl_dur, min_silent_dur)
>>> import numpy as np
>>> np.allclose(onsets, notmat_dict['onsets'])
True
```
(*Note that this test would return `False` if the onsets and offsets in the .not.mat 
annotation file had been modified, e.g., a user of the evsonganaly GUI had edited them,
after they were originally computed by the SegmentNotes.m function.*)

`evfuncs` is used to load annotations by  
['crowsetta'](https://github.com/NickleDave/crowsetta), 
a data-munging tool for building datasets of vocalizations 
that can be used to train machine learning models.
Two machine learning libraries that can use those datasets are: 
[`hybrid-vocal-classifier`](https://hybrid-vocal-classifier.readthedocs.io/en/latest/), 
and [`vak`](https://github.com/NickleDave/vak).

### Getting Help
Please feel free to raise an issue here:  
https://github.com/NickleDave/evfuncs/issues

### License
[BSD License](./LICENSE).

### Citation
Please cite this software as shown below.
To get the most up-to-date, automatically-generated citation, 
please click "Cite this repository" on the upper right side 
of the page.

bibtex:  
```
@software{Nicholson_evfuncs_2021,
author = {Nicholson, David},
doi = {10.5281/zenodo.4584209},
license = {BSD-3-Clause},
month = {3},
title = {{evfuncs}},
url = {https://github.com/NickleDave/evfuncs},
version = {0.3.2.post1},
year = {2021}
```

APA:  
```
Nicholson, D. (2021). evfuncs (Version 0.3.2.post1) [Computer software]. https://doi.org/10.5281/zenodo.4584209
```

### Build Status
[![Build Status](https://travis-ci.com/NickleDave/evfuncs.svg?branch=master)](https://travis-ci.com/NickleDave/evfuncs)


doc/CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.3.5] -- 2022-05-14
### Changed
- raise minimum required Python to 3.8, 
  to adhere to [NEP-29](https://numpy.org/neps/nep-0029-deprecation_policy.html)
  [#18](https://github.com/NickleDave/evfuncs/pull/18).
  Fixes [#17](https://github.com/NickleDave/evfuncs/issues/17).

## [0.3.4] -- 2022-03-17
### Changed
- change `evfuncs.load_notmat` so that the returned values for 
  `onsets` and `offsets` will always be `numpy.ndarray`, even when 
  the `.not.mat` only has a single annotated segment.
  [#16](https://github.com/NickleDave/evfuncs/pull/16)
  Fixes [#14](https://github.com/NickleDave/evfuncs/issues/14)

## [0.3.3] -- 2021-12-30
### Added
- added `CITATION.cff` file
  [#11](https://github.com/NickleDave/evfuncs/pull/11)

### Changed
- switch to using `flit` for development, to have PEP 621 metadata 
  and to not need `poetry` for `conda-forge` recipe, 
  that imposes limits on Python version
  [#12](https://github.com/NickleDave/evfuncs/pull/12)
- require minimum version of Python to be 3.7 
  [#13](https://github.com/NickleDave/evfuncs/pull/13)

## 0.3.2.post1 -- 2021-03-04
### Changed
- add metadata to pyproject.toml so that README is used as "long description" 
  and appears on PyPI
  [43c0742](https://github.com/NickleDave/evfuncs/commit/43c07428b8237b81bd0b80c61b5b27950ebb11bc)

## 0.3.2 -- 2021-03-04
### Changed
- switch to using GitHub Actions for continuous integration
  [62c1b89](https://github.com/NickleDave/evfuncs/commit/62c1b89a5bbcf12ac8c6929c79e46a9e966d7d47)

### Fixed
- change dependencies and required Python so they are not pinned to major version
  [65480ac](https://github.com/NickleDave/evfuncs/commit/65480ac3c50df6533284f57933134d9e34277086)

## 0.3.1
### Fixed
- fix development status within classifiers option of pyproject.toml
  [e7eee87](https://github.com/NickleDave/evfuncs/commit/e7eee870e3f1dc519acf5e6bd658b3c802a05841)

## 0.3.0
### Changed
- make it so that functions accept filenames as either string or `pathlib.Path`s
  [bd3494c](https://github.com/NickleDave/evfuncs/commit/bd3494c58bb32ce37b3cecc90d87469c075dca37)
- switch to using `poetry` for development
  [376e321](https://github.com/NickleDave/evfuncs/commit/376e3211ae6dc3e6a260c4da7967207f705634de)

## 0.2.1
### Fixed
- change `__init__.py` to import metadata from `__about__.py`, so that version 
is actually single-sourced (both __init__ and setup get it from same place)

## 0.2.0
### Added
- `__about__.py` file, to single-source version info and other metadata

### Changed
- `setup.py` uses `__about__.py` for version + other metadata

## Removed
- an extra `__init__.py` file that was in `src/`

## 0.1.1
- First real production version (now that I kind of mostly understand semantic versioning)
### Added
- Examples sections to most function docstrhttps://github.com/NickleDave/evfuncs/commit/bd3494c58bb32ce37b3cecc90d87469c075dca37ings

### Changed
- setup.py requires scipy>=1.2.0, fixes FutureWarnings raised by scipy.signal 
  (used in `evfuncs.bandpass_filtfilt`)

## 0.1.1a3
### Added
- actually add a changelog
- make a DOI
- add an image of Evren, because this is important

## 0.1.1a2
### Added
- Add evfuncs.segment_notes function and tests for it
- Add Matlab script `make_oracle_data_for_tests.m` that generates "oracle" data from 
actual functions written by Evren and saves in .mat files for comparison with output
from evfuncs

## 0.1.1a1
- Initial version after excising from hvc (https://github.com/NickleDave/hybrid-vocal-classifier/commits/0c50144d75e3a3205db82add8b48302edbbed511/hvc/evfuncs.py)
### Changed
- Change `evfuncs.readrecf` to consistently parse values the same way, e.g. sampling frequency always returned as an int
- Convert tests to Python unittest format (instead of using PyTest library)
- Change defaults for `evfuncs.smooth_data` so it acts like `SmoothData.m` (and passes those defaults to `evfuncs.bandpass_filtfilt`)
### Added
- Write README.md with usage


tests/test_evfuncs.py
"""
test evfuncs module
"""
import numpy as np
from scipy.io import loadmat

import evfuncs


def test_readrecf(rec_files):
    for rec_file in rec_files:
        rec_dict = evfuncs.readrecf(rec_file)
        assert 'header' in rec_dict
        assert type(rec_dict['header']) == list
        assert 'sample_freq' in rec_dict
        assert type(rec_dict['sample_freq']) == int or type(rec_dict['sample_freq']) == float
        assert 'num_channels' in rec_dict
        assert type(rec_dict['num_channels']) == int
        assert 'num_samples' in rec_dict
        assert type(rec_dict['num_samples']) == int
        assert 'iscatch' in rec_dict
        assert 'outfile' in rec_dict
        assert 'time_before' in rec_dict
        assert type(rec_dict['time_before']) == float
        assert 'time_after' in rec_dict
        assert type(rec_dict['time_after']) == float
        assert 'thresholds' in rec_dict
        assert type(rec_dict['thresholds']) == list
        assert all(
            [type(thresh) == float for thresh in rec_dict['thresholds']]
        )
        assert 'feedback_info' in rec_dict
        assert type(rec_dict['feedback_info']) == dict


def test_load_cbin(cbins):
    for cbin in cbins:
        dat, fs = evfuncs.load_cbin(cbin)
        assert type(dat) == np.ndarray
        assert dat.dtype == '>i2'  # should be big-endian 16 bit
        assert type(fs) == int


def test_load_notmat(notmats):
    for notmat in notmats:
        notmat_dict = evfuncs.load_notmat(notmat)
        assert type(notmat_dict) is dict
        assert 'onsets' in notmat_dict
        assert type(notmat_dict['onsets']) == np.ndarray
        assert notmat_dict['onsets'].dtype == float
        assert 'offsets' in notmat_dict
        assert type(notmat_dict['offsets']) == np.ndarray
        assert notmat_dict['offsets'].dtype == float
        assert 'labels' in notmat_dict
        assert type(notmat_dict['labels']) == str
        assert 'Fs' in notmat_dict
        assert type(notmat_dict['Fs']) == int
        assert 'fname' in notmat_dict
        assert type(notmat_dict['fname']) == str
        assert 'min_int' in notmat_dict
        assert type(notmat_dict['min_int']) == int
        assert 'min_dur' in notmat_dict
        assert type(notmat_dict['min_dur']) == int
        assert 'threshold' in notmat_dict
        assert type(notmat_dict['threshold']) == int
        assert 'sm_win' in notmat_dict
        assert type(notmat_dict['sm_win']) == int


def test_load_notmat_single_annotated_segment(notmat_with_single_annotated_segment):
    notmat_dict = evfuncs.load_notmat(notmat_with_single_annotated_segment)
    assert type(notmat_dict) is dict
    assert 'onsets' in notmat_dict
    assert type(notmat_dict['onsets']) == np.ndarray
    assert notmat_dict['onsets'].dtype == float
    assert 'offsets' in notmat_dict
    assert type(notmat_dict['offsets']) == np.ndarray
    assert notmat_dict['offsets'].dtype == float
    assert 'labels' in notmat_dict
    assert type(notmat_dict['labels']) == str
    assert 'Fs' in notmat_dict
    assert type(notmat_dict['Fs']) == int
    assert 'fname' in notmat_dict
    assert type(notmat_dict['fname']) == str
    assert 'min_int' in notmat_dict
    assert type(notmat_dict['min_int']) == int
    assert 'min_dur' in notmat_dict
    assert type(notmat_dict['min_dur']) == int
    assert 'threshold' in notmat_dict
    assert type(notmat_dict['threshold']) == int
    assert 'sm_win' in notmat_dict
    assert type(notmat_dict['sm_win']) == int


def test_bandpass_filtfilt_works(cbins, filtsong_mat_files):
    for cbin, filtsong_mat_file in zip(cbins, filtsong_mat_files):
        dat, fs = evfuncs.load_cbin(cbin)
        filtsong = evfuncs.bandpass_filtfilt(dat, fs)
        assert type(filtsong) == np.ndarray
        filtsong_mat = loadmat(filtsong_mat_file)
        filtsong_mat = np.squeeze(filtsong_mat['filtsong'])
        assert np.allclose(filtsong, filtsong_mat)


def test_smooth_data(cbins, smooth_data_mat_files):
    for cbin, smooth_data_mat_file in zip(cbins, smooth_data_mat_files):
        dat, fs = evfuncs.load_cbin(cbin)
        smoothed = evfuncs.smooth_data(dat, fs, freq_cutoffs=None)
        smoothed_500_10k = evfuncs.smooth_data(dat, fs,
                                               freq_cutoffs=(500, 10000))
        assert type(smoothed) == np.ndarray
        assert type(smoothed_500_10k) == np.ndarray
        assert not np.all(np.equal(smoothed, smoothed_500_10k))
        smooth_data_mat = loadmat(smooth_data_mat_file)
        smooth_data_mat = np.squeeze(smooth_data_mat['sm'])
        assert np.allclose(smoothed_500_10k, smooth_data_mat)


def test_segment_song(cbins, notmats, segment_mats):
    for cbin, notmat, segment_mat in zip(cbins, notmats, segment_mats):
        dat, fs = evfuncs.load_cbin(cbin)
        smooth = evfuncs.smooth_data(dat, fs)
        nmd = evfuncs.load_notmat(notmat)
        min_syl_dur = nmd['min_dur'] / 1000
        min_silent_dur = nmd['min_int'] / 1000
        threshold = nmd['threshold']
        onsets, offsets = evfuncs.segment_song(smooth, fs,
                                               threshold, min_syl_dur, min_silent_dur)
        segment_dict = loadmat(segment_mat, squeeze_me=True)
        onsets_mat = segment_dict['onsets']
        offsets_mat = segment_dict['offsets']
        # set tolerances for numpy.allclose check.
        # difference np.abs(offsets - offsets_mat) is usually ~0.00003125...
        # We just don't want error to be larger than a millisecond
        # By trial and error, I find that these vals for tolerance result in
        # about that ceiling
        atol = 0.0005
        rtol = 0.00001
        # i.e., 0.0005 + 0.00001 * some_onsets_or_offset_array ~ [0.0005, 0.0005, ...]
        assert np.allclose(onsets, onsets_mat, rtol, atol)
        assert np.allclose(offsets, offsets_mat, rtol, atol)


tests/fixtures/files.py
import pytest


@pytest.fixture
def cbins(gy6or6_032312_subset_root):
    return sorted(gy6or6_032312_subset_root.glob('*.cbin'))


@pytest.fixture
def notmats(gy6or6_032312_subset_root):
    return sorted(gy6or6_032312_subset_root.glob('*.not.mat'))


@pytest.fixture
def rec_files(gy6or6_032312_subset_root):
    return sorted(gy6or6_032312_subset_root.glob('*.rec'))


@pytest.fixture
def filtsong_mat_files(gy6or6_032312_subset_root):
    return sorted(gy6or6_032312_subset_root.glob('*filtsong*.mat'))


@pytest.fixture
def smooth_data_mat_files(gy6or6_032312_subset_root):
    return sorted(gy6or6_032312_subset_root.glob('*smooth_data*.mat'))


@pytest.fixture
def segment_mats(gy6or6_032312_subset_root):
    return sorted(gy6or6_032312_subset_root.glob('*unedited_SegmentNotes_output.mat'))


@pytest.fixture
def notmat_with_single_annotated_segment(data_for_tests_root):
    return data_for_tests_root / 'or60yw70-song-edited-to-have-single-segment' / 'or60yw70_300912_0725.437.cbin.not.mat'


tests/fixtures/paths.py
import pathlib

import pytest


HERE = pathlib.Path(__file__).parent
DATA_FOR_TESTS_ROOT = HERE / '..' / 'data_for_tests'


@pytest.fixture
def data_for_tests_root():
    return DATA_FOR_TESTS_ROOT


@pytest.fixture
def gy6or6_032312_subset_root(data_for_tests_root):
    return data_for_tests_root / 'gy6or6_032312_subset'


src/evfuncs/__about__.py
import os.path

__all__ = [
    "__title__",
    "__summary__",
    "__uri__",
    "__version__",
    "__commit__",
    "__author__",
    "__email__",
    "__license__",
    "__copyright__",
]


try:
    base_dir = os.path.dirname(os.path.abspath(__file__))
except NameError:
    base_dir = None


__title__ = "evfuncs"
__summary__ = "Functions for working with files created by the EvTAF program and the evsonganaly GUI"
__uri__ = "https://github.com/NickleDave/evfuncs"

__version__ = '0.3.5'

if base_dir is not None and os.path.exists(os.path.join(base_dir, ".commit")):
    with open(os.path.join(base_dir, ".commit")) as fp:
        __commit__ = fp.read().strip()
else:
    __commit__ = None

__author__ = "David Nicholson"
__email__ = "nicholdav@gmail.com"

__license__ = "BSD"
__copyright__ = "2018-present %s" % __author__


src/evfuncs/evfuncs.py
"""
ev_funcs
Python implementations of functions used with EvTAF and evsonganaly.m
"""
from pathlib import Path

import numpy as np
from scipy.io import loadmat
import scipy.signal


def readrecf(filename):
    """reads .rec files output by EvTAF
    and returns rec_dict with (approximately)
    the same structure as the rec struct
    returned by readrec.m

    Parameters
    ----------
    filename : str, Path
        name of .rec file, can include path

    Returns
    -------
    rec_dict : dict
        with following key, value pairs
            header : str
                header from .rec file
            time_before : float
                time in seconds of additional audio added to file by EvTAF
                from before sound amplitude went above threshold for recording
            time_after : float
                time in seconds of additional audio added to file by EvTAF
                after sound amplitude dropped below threshold for recording
            iscatch : bool
                list of whether each occurrence of auditory feedback triggered
                was a 'catch trial' (where feedback would have been triggered but
                was withheld so behavior on that trial could be measured)
            num_channels : int
                number of channels from DAQ board recored by EvTAF
            sample_freq : int
                sampling frequency of audio file associated with this .rec file, in Hertz
            num_samples : int
                number of audio samples in file
            outfile : str
                name of audio file played as auditory feedback (if there was any)
            thresholds : int
                thresholds used by ring counter to determine whether to trigger auditory feedback
            feedback_info : dict
                maps feedback type to time it occurred in milliseconds

    Examples
    --------
    >>> recf = 'gy6or6_baseline_230312_0808.138.rec'
    >>> rec_dict = readrecf(recf)
    >>> num_samples = rec_dict['num_samples'}
    >>> sample_freq = rec_dict['sample_freq'}
    >>> print(f"file duration in seconds: {num_samples / sample_freq:.3f}")
    file duration in seconds: 12.305
    """
    filename = Path(filename)

    rec_dict = {}
    with filename.open('r') as recfile:
        line_tmp = ""
        while 1:
            if line_tmp == "":
                line = recfile.readline()
            else:
                line = line_tmp
                line_tmp = ""
                
            if line == "":  # if End Of File
                break
            elif line == "\n":  # if blank line
                continue
            elif "Catch" in line:
                ind = line.find('=')
                rec_dict['iscatch'] = line[ind+1:]
            elif "Chans" in line:
                ind = line.find('=')
                rec_dict['num_channels'] = int(line[ind+1:])
            elif "ADFREQ" in line:
                ind = line.find('=')
                try:
                    rec_dict['sample_freq'] = int(line[ind+1:])
                except ValueError:
                    # if written with scientific notation
                    # first parse as float, then cast to int
                    sample_freq_float = float(line[ind+1:])
                    try:
                        rec_dict['sample_freq'] = int(sample_freq_float)
                    except ValueError:
                        raise ValueError("Couldn't convert following value for "
                                         "ADFREQ in .rec file {} to an integer: "
                                         "{}".format(filename,
                                                     sample_freq_float))
            elif "Samples" in line:
                ind = line.find('=')
                rec_dict['num_samples'] = int(line[ind+1:])
            elif "T After" in line or "T AFTER" in line:
                ind = line.find('=')
                rec_dict['time_after'] = float(line[ind+1:])
            elif "T Before" in line or "T BEFORE" in line:
                ind = line.find('=')
                rec_dict['time_before'] = float(line[ind+1:])
            elif "Output Sound File" in line:
                ind = line.find('=')
                rec_dict['outfile'] = line[ind+1:]
            elif "Thresholds" in line or "THRESHOLDS" in line:
                th_list = []
                while 1:
                    line = recfile.readline()
                    if line == "":
                        break
                    try:
                        th_list.append(float(line))
                    except ValueError:  # because we reached next section
                        line_tmp = line
                        break
                rec_dict['thresholds'] = th_list
                if line == "":
                    break
            elif "Feedback information" in line:
                fb_dict = {}
                while 1:
                    line = recfile.readline()
                    if line == "":
                        break
                    elif line == "\n":
                        continue
                    ind = line.find("msec")
                    time = float(line[:ind-1])
                    ind = line.find(":")
                    fb_type = line[ind+2:]
                    fb_dict[time] = fb_type
                rec_dict['feedback_info'] = fb_dict
                if line == "":
                    break
            elif "File created" in line:
                header = [line]
                for counter in range(4):
                    line = recfile.readline()
                    header.append(line)
                rec_dict['header']=header
    return rec_dict


def load_cbin(filename, channel=0):
    """loads .cbin files output by EvTAF.
    
    Parameters
    ----------
    filename : str
        name of .cbin file, can include path
    channel : int
        Channel in file to load. Default is 0.

    Returns
    -------
    data : numpy.ndarray
        1-d vector of 16-bit signed integers
    sample_freq : int or float
        sampling frequency in Hz. Typically 32000.

    Examples
    --------
    >>> cbin_filename = 'gy6or6_baseline_230312_0808.138.cbin'
    >>> data, sample_freq = load_cbin(cbin_filename)
    >>> data
    array([-230, -223, -235, ...,   34,   36,   26], dtype=int16)
    """
    filename = Path(filename)

    # .cbin files are big endian, 16 bit signed int, hence dtype=">i2" below
    data = np.fromfile(filename, dtype=">i2")
    recfile = filename.parent.joinpath(filename.stem + '.rec')
    rec_dict = readrecf(recfile)
    data = data[channel::rec_dict['num_channels']]  # step by number of channels
    sample_freq = rec_dict['sample_freq']
    return data, sample_freq


def load_notmat(filename):
    """loads .not.mat files created by evsonganaly (Matlab GUI for labeling song)

    Parameters
    ----------
    filename : str
        name of .not.mat file, can include path

    Returns
    -------
    notmat_dict : dict
        variables from .not.mat files

    Examples
    --------
    >>> a_notmat = 'gy6or6_baseline_230312_0808.138.cbin.not.mat'
    >>> notmat_dict = load_notmat(a_notmat)
    >>> notmat_dict.keys()
    dict_keys(['__header__', '__version__', '__globals__', 'Fs', 'fname', 'labels',
    'onsets', 'offsets', 'min_int', 'min_dur', 'threshold', 'sm_win'])

    Notes
    -----
    Basically a wrapper around `scipy.io.loadmat`. Calls `loadmat` with `squeeze_me=True`
    to remove extra dimensions from arrays that `loadmat` parser sometimes adds.

    Also note that **onsets and offsets from .not.mat files are in milliseconds**.
    The GUI `evsonganaly` saves onsets and offsets in these units,
    and we avoid converting them here for consistency and interoperability
    with Matlab code.
    """
    filename = Path(filename)

    # have to cast to str and call endswith because 'ext' from Path will just be .mat
    if str(filename).endswith(".not.mat"):
        pass
    elif str(filename).endswith("cbin"):
        filename = filename.parent.joinpath(filename.name + ".not.mat")
    else:
        ext = filename.suffix
        raise ValueError(
            f"Filename should have extension .cbin.not.mat or .cbin but extension was: {ext}"
        )
    notmat_dict = loadmat(filename, squeeze_me=True)
    # ensure that onsets and offsets are always arrays, not scalar
    for key in ('onsets', 'offsets'):
        if np.isscalar(notmat_dict[key]):  # `squeeze_me` makes them a ``float``, this will be True in that case
            value = np.array(notmat_dict[key])[np.newaxis]  # ``np.newaxis`` ensures 1-d array with shape (1,)
            notmat_dict[key] = value
    return notmat_dict


def bandpass_filtfilt(rawsong, samp_freq, freq_cutoffs=(500, 10000)):
    """filter song audio with band pass filter, then perform zero-phase
    filtering with filtfilt function

    Parameters
    ----------
    rawsong : ndarray
        audio
    samp_freq : int
        sampling frequency
    freq_cutoffs : list
        2 elements long, cutoff frequencies for bandpass filter.
        Default is [500, 10000].

    Returns
    -------
    filtsong : ndarray
    """
    if freq_cutoffs[0] <= 0:
        raise ValueError('Low frequency cutoff {} is invalid, '
                         'must be greater than zero.'
                         .format(freq_cutoffs[0]))

    Nyquist_rate = samp_freq / 2
    if freq_cutoffs[1] >= Nyquist_rate:
        raise ValueError('High frequency cutoff {} is invalid, '
                         'must be less than Nyquist rate, {}.'
                         .format(freq_cutoffs[1], Nyquist_rate))

    if rawsong.shape[-1] < 387:
        numtaps = 64
    elif rawsong.shape[-1] < 771:
        numtaps = 128
    elif rawsong.shape[-1] < 1539:
        numtaps = 256
    else:
        numtaps = 512

    cutoffs = np.asarray([freq_cutoffs[0] / Nyquist_rate,
                          freq_cutoffs[1] / Nyquist_rate])
    # code on which this is based, bandpass_filtfilt.m, says it uses Hann(ing)
    # window to design filter, but default for matlab's fir1
    # is actually Hamming
    # note that first parameter for scipy.signal.firwin is filter *length*
    # whereas argument to matlab's fir1 is filter *order*
    # for linear FIR, filter length is filter order + 1
    b = scipy.signal.firwin(numtaps + 1, cutoffs, pass_zero=False)
    a = np.zeros((numtaps+1,))
    a[0] = 1  # make an "all-zero filter"
    padlen = np.max((b.shape[-1] - 1, a.shape[-1] - 1))
    filtsong = scipy.signal.filtfilt(b, a, rawsong, padlen=padlen)
    return filtsong


def smooth_data(rawsong, samp_freq, freq_cutoffs=(500, 10000), smooth_win=2):
    """filter raw audio and smooth signal
    used to calculate amplitude.

    Parameters
    ----------
    rawsong : ndarray
        1-d numpy array, "raw" voltage waveform from microphone
    samp_freq : int
        sampling frequency
    freq_cutoffs: list
        two-element list of integers, [low freq., high freq.]
        bandpass filter applied with this list defining pass band.
        If None, in which case bandpass filter is not applied.
    smooth_win : integer
        size of smoothing window in milliseconds. Default is 2.

    Returns
    -------
    smooth : ndarray
        1-d numpy array, smoothed waveform

    Applies a bandpass filter with the frequency cutoffs in spect_params,
    then rectifies the signal by squaring, and lastly smooths by taking
    the average within a window of size sm_win.
    This is a very literal translation from the Matlab function SmoothData.m
    by Evren Tumer. Uses the Thomas-Santana algorithm.
    """
    if freq_cutoffs is None:
        # then don't do bandpass_filtfilt
        filtsong = rawsong
    else:
        filtsong = bandpass_filtfilt(rawsong, samp_freq, freq_cutoffs)

    squared_song = np.power(filtsong, 2)
    len = np.round(samp_freq * smooth_win / 1000).astype(int)
    h = np.ones((len,)) / len
    smooth = np.convolve(squared_song, h)
    offset = round((smooth.shape[-1] - filtsong.shape[-1]) / 2)
    smooth = smooth[offset:filtsong.shape[-1] + offset]
    return smooth


def segment_song(smooth, samp_freq, threshold=5000, min_syl_dur=0.02,
                 min_silent_dur=0.002, return_Hz=False):
    """segment audio file of birdsong into syllables

    Parameters
    ----------
    smooth : np.ndarray
        Smoothed audio waveform, returned by evfuncs.smooth_data
    samp_freq : int
        Sampling frequency at which audio was recorded. Returned by
        evfuncs.load_cbin.
    threshold : int
        value above which amplitude is considered part of a segment.
        Default is 5000.
    min_syl_dur : float
        minimum duration of a segment.
        In .not.mat files saved by evsonganaly, this value is called "min_dur"
        Default is 0.02, i.e. 20 ms.
    min_silent_dur : float
        minimum duration of silent gap between segment.
        In .not.mat files saved by evsonganaly, this value is called "min_int"
        Default is 0.002, i.e. 2 ms.
    return_Hz : bool
        if True, returns the onsets and offsets in units of Hz.
        Default is False.

    Returns
    -------
    onsets_s : np.ndarray
        Onset times of syllables in song, in seconds.
    offsets_s : np.ndarray
        Offset times of syllables in song, in seconds.
    onsets_Hz : np.ndarray
        Onset times of syllables in song, in seconds.
        Only returned if "return_Hz" is True.
    offsets_Hz : np.ndarray
        Offset times of syllables in song, in seconds.
        Only returned if "return_Hz" is True.

    Equivalent to SegmentNotes.m function that is part of evsonganaly GUI.
    """
    above_th = smooth > threshold
    h = [1, -1]
    # convolving with h causes:
    # +1 whenever above_th changes from 0 to 1
    # and -1 whenever above_th changes from 1 to 0
    above_th_convoluted = np.convolve(h, above_th)

    # always get in units of Hz first, then convert to s
    onsets_Hz = np.where(above_th_convoluted > 0)[0]
    offsets_Hz = np.where(above_th_convoluted < 0)[0]
    onsets_s = onsets_Hz / samp_freq
    offsets_s = offsets_Hz / samp_freq

    if onsets_s.shape[0] < 1 or offsets_s.shape[0] < 1:
        return None, None  # because no onsets or offsets in this file

    # get rid of silent intervals that are shorter than min_silent_dur
    silent_gap_durs = onsets_s[1:] - offsets_s[:-1]  # duration of silent gaps
    keep_these = np.nonzero(silent_gap_durs > min_silent_dur)
    onsets_s = np.concatenate(
        (onsets_s[0, np.newaxis], onsets_s[1:][keep_these]))
    offsets_s = np.concatenate(
        (offsets_s[:-1][keep_these], offsets_s[-1, np.newaxis]))
    if return_Hz:
        onsets_Hz = np.concatenate(
            (onsets_Hz[0, np.newaxis], onsets_Hz[1:][keep_these]))
        offsets_Hz = np.concatenate(
            (offsets_Hz[:-1][keep_these], offsets_Hz[-1, np.newaxis]))

    # eliminate syllables with duration shorter than min_syl_dur
    syl_durs = offsets_s - onsets_s
    keep_these = np.nonzero(syl_durs > min_syl_dur)
    onsets_s = onsets_s[keep_these]
    offsets_s = offsets_s[keep_these]
    if return_Hz:
        onsets_Hz = onsets_Hz[keep_these]
        offsets_Hz = offsets_Hz[keep_these]

    if return_Hz:
        return onsets_s, offsets_s, onsets_Hz, offsets_Hz
    else:
        return onsets_s, offsets_s
